tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+    {"wasi_unstable", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_unstable", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_unstable", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_unstable", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_unstable", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+};
+
+bool run(int argc, char** argv)
+{
+    uvwasi_options_t options = {
+        3,           // sizeof fd_table
+        0, nullptr,  // NOTE: no remappings
+        static_cast<uvwasi_size_t>(argc), const_cast<const char**>(argv),
+        nullptr,  // TODO: support env
+        0, 1, 2,
+        nullptr,  // NOTE: no special allocator
+    };
+
+    uvwasi_errno_t err = uvwasi_init(&state, &options);
+    if (err != UVWASI_ESUCCESS)
+    {
+        std::cerr << "Failed to initialise UVWASI: " << uvwasi_embedder_err_code_to_string(err)
+                  << std::endl;
+        return false;
+    }
+
+    if (!std::filesystem::is_regular_file(argv[0]))
+    {
+        std::cerr << "File does not exists or is irregular: " << argv[0] << std::endl;
+        return false;
+    }
+
+    std::ifstream wasm_file{argv[0]};
+    if (!wasm_file)
+    {
+        std::cerr << "Failed to open file: " << argv[0] << std::endl;
+        return false;
+    }
+    const auto wasm_binary =
+        fizzy::bytes(std::istreambuf_iterator<char>{wasm_file}, std::istreambuf_iterator<char>{});
+
+    auto module = fizzy::parse(wasm_binary);
+    auto imports = fizzy::resolve_imported_functions(module, wasi_functions);
+    auto instance = fizzy::instantiate(std::move(module), std::move(imports));
+    assert(instance != nullptr);
+
+    const auto start_function = fizzy::find_exported_function(instance->module, "_start");
+    if (!start_function.has_value())
+    {
+        std::cerr << "File is not WASI compatible (_start not found)" << std::endl;
+        return false;
+    }
+
+    // Manually validate type signature here
+    // TODO: do this in find_exported_function
+    if (instance->module.get_function_type(*start_function) != fizzy::FuncType{})
+    {
+        std::cerr << "File is not WASI compatible (_start has invalid signature)" << std::endl;
+        return false;
+    }
+
+    if (!fizzy::find_exported_memory(*instance, "memory").has_value())
+    {
+        std::cerr << "File is not WASI compatible (no memory exported)" << std::endl;
+        return false;
+    }
+
+    const auto [trapped, result_stack] = fizzy::execute(*instance, *start_function, {});
+    assert(result_stack.empty());
```

Comment:
I think we return stack of trapped function? Then this will fail, should assert only if not trapped.

Reply:
Hmm, good point. Will just move the assert.